syntax = "proto2";

package jraft;

option cc_generic_services = true;


message RequestVoteRequest {
    required int64 term = 1;
    required int64 peerId = 2;
    required int64 candidateId = 3;
    required int64 lastLogIndex = 4;
    required int64 lastLogTerm = 5;
}

message RequestVoteResponse {
    required int64 term = 1;
    required bool voteGranted = 2;
    required int64 peerId = 3;
}

message AppendEntriesRequest {
    required int64 term = 1;
    required int64 peerId = 2;
    required int64 prevLogIndex = 3;
    required int64 prevLogTerm = 4;
    repeated string entries = 5;
    required int64 commitedIndex = 6;
}

message AppendEntriesResponse {
    required int64 prevLogIndex = 1;
    required int64 prevLogTerm = 2;
    required bool success = 3;
    required int64 peerId = 4;
}

service RaftService {
    rpc PreVote(RequestVoteRequest) returns (RequestVoteResponse);
    rpc RequestVote(RequestVoteRequest) returns (RequestVoteResponse);
    rpc AppendEntries(AppendEntriesRequest) returns (AppendEntriesResponse);
}
